src/VisualStudio/Core/Def/Implementation/CodeLens/CodeLensCallbackListener.cs

```diff
@@ -0,0 +1,170 @@
+        public CodeLensCallbackListener(
+            IThreadingContext threadingContext,
+            SVsServiceProvider serviceProvider,
+            VisualStudioWorkspaceImpl workspace)
+        {
+            _threadingContext = threadingContext;
+            _serviceProvider = serviceProvider;
+            _workspace = workspace;
+        }
+
+        public async Task<string> GetHostGroupIdAsync(CancellationToken cancellationToken)
+        {
+            var client = await _workspace.TryGetRemoteHostClientAsync(cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                // exception is handled by code lens engine
+                throw new InvalidOperationException("remote host doesn't exist");
+            }
+
+            return client.ClientId;
+        }
+
+        public List<Guid> GetDocumentId(Guid projectGuid, string filePath, CancellationToken cancellationToken)
+        {
+            if (TryGetDocument(_workspace.CurrentSolution, projectGuid, filePath, out var document))
+            {
+                var documentId = document.Id;
+                return new List<Guid>(2) { documentId.ProjectId.Id, documentId.Id };
+            }
+
+            return null;
+        }
+
+        public async Task<ReferenceCount> GetReferenceCountAsync(
+            CodeLensDescriptor descriptor, CodeLensDescriptorContext descriptorContext, CancellationToken cancellationToken)
+        {
+            var solution = _workspace.CurrentSolution;
+            if (!TryGetDocument(solution, descriptor.ProjectGuid, descriptor.FilePath, out var document))
+            {
+                return null;
+            }
+
+            var maxSearchResults = await GetMaxResultCapAsync(cancellationToken).ConfigureAwait(false);
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(descriptorContext.ApplicableSpan.Value.ToTextSpan());
+
+            var service = _workspace.Services.GetService<ICodeLensReferencesService>();
+            return await service.GetReferenceCountAsync(solution, document.Id, node, maxSearchResults, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<IEnumerable<ReferenceLocationDescriptor>> FindReferenceLocationsAsync(
+            CodeLensDescriptor descriptor, CodeLensDescriptorContext descriptorContext, CancellationToken cancellationToken)
+        {
+            var solution = _workspace.CurrentSolution;
+            if (!TryGetDocument(solution, descriptor.ProjectGuid, descriptor.FilePath, out var document))
+            {
+                return null;
+            }
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(descriptorContext.ApplicableSpan.Value.ToTextSpan());
+
+            var service = _workspace.Services.GetService<ICodeLensReferencesService>();
```

Comment:
duplicated code between `GetReferenceCountAsync` and `FindReferenceLocationsAsync`